controllers/subscriptions_test.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2021 Red Hat OpenShift Data Foundation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"testing"
+
+	operatorv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	odfv1alpha1 "github.com/red-hat-data-services/odf-operator/api/v1alpha1"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestEnsureSubscription(t *testing.T) {
+
+	testCases := []struct {
+		kind                     odfv1alpha1.StorageKind
+		subscriptionAlreadyExist bool
+		expectedSubscription     bool
+	}{
+
+		{
+			kind:                     odfv1alpha1.StorageCluster,
+			subscriptionAlreadyExist: false,
+			expectedSubscription:     false,
+		},
+		{
+			kind:                     odfv1alpha1.FlashSystemCluster,
+			subscriptionAlreadyExist: false,
+			expectedSubscription:     true,
+		},
+		{
+			kind:                     odfv1alpha1.FlashSystemCluster,
+			subscriptionAlreadyExist: true,
+			expectedSubscription:     true,
+		},
+	}
+
+	for _, tc := range testCases {
+		fakeReconciler, fakeStorageSystem := GetFakeStorageSystemReconciler()
+		err := operatorv1alpha1.AddToScheme(fakeReconciler.Scheme)
+		assert.NoError(t, err)
+
+		if tc.kind == odfv1alpha1.FlashSystemCluster {
+			fakeStorageSystem.Spec.Kind = odfv1alpha1.FlashSystemCluster
+		}
+
+		if tc.subscriptionAlreadyExist {
+			subscription := GetFlashSystemClusterSubscription(fakeStorageSystem)
+			subscription.Spec.Channel = "fake-channel"
+			err := fakeReconciler.Client.Create(context.TODO(), subscription)
+			assert.NoError(t, err)
+		}
+
+		err = fakeReconciler.ensureSubscription(fakeStorageSystem, fakeReconciler.Log)
+		assert.NoError(t, err)
+
+		existingSubscriptions := &operatorv1alpha1.SubscriptionList{}
+		err = fakeReconciler.Client.List(context.TODO(), existingSubscriptions)
+		assert.NoError(t, err)
+
+		if !tc.expectedSubscription {
+			assert.Equal(t, 0, len(existingSubscriptions.Items))
+		} else {
+			assert.Equal(t, 1, len(existingSubscriptions.Items))
+			assert.Equal(t, "ibm-operator", existingSubscriptions.Items[0].Spec.Package)
+			assert.Equal(t, "alpha", existingSubscriptions.Items[0].Spec.Channel)
```

Comment:
This is fine for now, but I would like to see this be a little less hardcoded. Add a comment like `//TODO: Parameterize this`.

Reply:
Done